pilot/pkg/serviceregistry/aggregate/controller.go

```diff
@@ -358,12 +366,35 @@ func (c *Controller) AppendWorkloadHandler(f func(*model.WorkloadInstance, model
 	return nil
 }
 
-// GetIstioServiceAccounts implements model.ServiceAccounts operation
+// GetIstioServiceAccounts implements model.ServiceAccounts operation.
+// The returned list contains all SPIFFE based identities that backs the service. For example,
+// - { "spiffe://cluster.local/bar@iam.gserviceaccount.com"}; when annotation is used on corresponding workloads.
+// - { "spiffe://cluster.local/ns/default/sa/foo" }; normal kubernetes cases
+// - { "spiffe://cluster.local/ns/default/sa/foo", "spiffe://trust-domain-alias/ns/default/sa/foo" };
+//   if the trust domain alias is configured.
 func (c *Controller) GetIstioServiceAccounts(svc *model.Service, ports []int) []string {
+	out := map[string]struct{}{}
 	for _, r := range c.GetRegistries() {
-		if svcAccounts := r.GetIstioServiceAccounts(svc, ports); svcAccounts != nil {
-			return svcAccounts
+		svcAccounts := r.GetIstioServiceAccounts(svc, ports)
+		for _, sa := range svcAccounts {
+			out[sa] = struct{}{}
 		}
 	}
-	return nil
+	result := []string{}
+	for k := range out {
+		result = append(result, k)
+	}
+	tds := []string{}
+	if c.meshHolder != nil {
+		mesh := c.meshHolder.Mesh()
+		if mesh != nil {
+			tds = mesh.TrustDomainAliases
+		}
+	}
+	expanded := spiffe.ExpandWithTrustDomains(result, tds)
+	result = []string{}
+	for k := range expanded {
+		result = append(result, k)
+	}
+	return result
```

Comment:
does order matter here? order will not be deterministic due to the iteration over a map. Will this in turn lead to non-deterministic ordering in CDS and connection draining?

Reply:
it should not matter AFAIK.  the push trigger is somewhere in the eds update code to compare whether service account set are the same (before expansion). But let me sort just in case (most time the list should be short, one or up to handful service account and single trust domain, so should not be too much complexity in runtime).